shell/renderer/api/electron_api_context_bridge.cc

```diff
@@ -810,13 +854,137 @@ bool OverrideGlobalPropertyFromIsolatedWorld(
-  auto* frame = render_frame->GetWebFrame();
-  CHECK(frame);
-  v8::Local<v8::Context> main_context = frame->MainWorldScriptContext();
-  return isolate->GetCurrentContext() == main_context;
+  v8::Local<v8::Context> source_context = isolate->GetCurrentContext();
+  auto* ec = blink::ExecutionContext::From(source_context);
+  if (ec->IsWindow()) {
+    auto* render_frame = GetRenderFrame(source_context->Global());
+    CHECK(render_frame);
+    auto* frame = render_frame->GetWebFrame();
+    CHECK(frame);
+    v8::Local<v8::Context> main_context = frame->MainWorldScriptContext();
+    return source_context == main_context;
+  } else if (ec->IsShadowRealmGlobalScope()) {
+    return false;
+  } else if (ec->IsServiceWorkerGlobalScope()) {
+    return true;
+  } else {
+    NOTREACHED();
+  }
+}
+
+// Clones a value into the target context.
+v8::MaybeLocal<v8::Value> CloneValueToContext(
+    v8::Isolate* isolate,
+    v8::Local<v8::Value> value,
+    v8::Local<v8::Context> target_context,
+    std::string& error_message) {
+  // Objects with prototype chains need to be protected and thus cloned.
+  // Primitive values are fine as is.
+  if (!value->IsObject()) {
+    return value;
+  }
+
+  v8::Local<v8::Object> object_value = value.As<v8::Object>();
+  v8::Local<v8::Context> creation_context =
+      object_value->GetCreationContextChecked();
+
+  // Value created in the same context are safe.
+  if (target_context == creation_context) {
+    return object_value;
+  }
+
+  // Attempt to clone value.
+  v8::MaybeLocal<v8::Value> maybe_result;
+  {
+    v8::TryCatch try_catch(isolate);
+    context_bridge::ObjectCache object_cache;
+    maybe_result =
+        PassValueToOtherContext(creation_context, target_context, object_value,
+                                creation_context->Global(), &object_cache,
+                                false, 0, BridgeErrorTarget::kSource);
+    if (try_catch.HasCaught()) {
+      v8::String::Utf8Value utf8(isolate, try_catch.Exception());
+      error_message = *utf8 ? *utf8 : "Unknown error cloning result";
+    }
+  }
+
+  return maybe_result;
+}
+
+// Evaluate a script in the target world ID. The script is executed
+// synchronously and clones the result into the calling context.
+v8::Local<v8::Value> EvaluateInWorld(v8::Isolate* isolate,
+                                     const int world_id,
+                                     const std::string& source,
+                                     gin_helper::Arguments* args) {
+  v8::Local<v8::Context> source_context = isolate->GetCurrentContext();
+  v8::Context::Scope source_scope(source_context);
+
+  // Get the target context
+  v8::MaybeLocal<v8::Context> maybe_target_context =
+      GetTargetContext(isolate, world_id);
+  v8::Local<v8::Context> target_context;
+  if (!maybe_target_context.ToLocal(&target_context)) {
+    isolate->ThrowException(v8::Exception::Error(
+        gin::StringToV8(isolate, "Unknown error")));  // TODO
```

Comment:
Maybe some kind of error message to indicate target context not found along with the world id?
```
   isolate->ThrowException(v8::Exception::Error(
       gin::StringToV8(isolate, "Failed to retrieve target context for world ID: " + std::to_string(world_id))));
```